api/src/transactions.rs

```diff
@@ -563,26 +563,51 @@ impl TransactionsApi {
-                    ident_str!("balance").into(),
-                    vec![AptosCoinType::type_tag()],
-                    vec![signed_transaction.sender().to_vec()],
-                    context.node_config.api.max_gas_view_function,
-                );
-                let values = output.values.map_err(|err| {
-                    SubmitTransactionError::bad_request_with_code_no_info(
-                        err,
-                        AptosErrorCode::InvalidInput,
-                    )
-                })?;
-                let balance: u64 = bcs::from_bytes(&values[0]).map_err(|err| {
-                    SubmitTransactionError::bad_request_with_code_no_info(
-                        err,
-                        AptosErrorCode::InvalidInput,
-                    )
-                })?;
+
+                // To determine who to check the amount of, need to determine if there's a fee payer
+                let fee_payer_address = if let Some(address) =
+                    signed_transaction.authenticator_ref().fee_payer_address()
+                {
+                    if address.eq(&AccountAddress::ZERO) {
+                        // This means we don't know who the fee payer is, but there is one.
+                        // In this case, we'll have to estimate based on the max gas amount, though
+                        // it may fail
+                        None
+                    } else {
+                        // If it's defined, then use that address
+                        Some(address)
+                    }
+                } else {
+                    // If no fee payer, use the sender
+                    Some(signed_transaction.sender())
+                };
+
+                let balance: u64 = if let Some(fee_payer_address) = fee_payer_address {
+                    let output = AptosVM::execute_view_function(
+                        &state_view,
+                        ModuleId::new(AccountAddress::ONE, ident_str!("coin").into()),
+                        ident_str!("balance").into(),
+                        vec![AptosCoinType::type_tag()],
+                        vec![fee_payer_address.to_vec()],
+                        context.node_config.api.max_gas_view_function,
+                    );
+                    let values = output.values.map_err(|err| {
+                        SubmitTransactionError::bad_request_with_code_no_info(
+                            err,
+                            AptosErrorCode::InvalidInput,
+                        )
+                    })?;
+                    bcs::from_bytes(&values[0]).map_err(|err| {
+                        SubmitTransactionError::bad_request_with_code_no_info(
+                            err,
+                            AptosErrorCode::InvalidInput,
+                        )
+                    })?
+                } else {
+                    // If we don't know who's paying, we just have to put in an assumed number, right
```

Comment:
I wonder if it's possible to leave it as 0x0 and just "fund" 0x0 by send 2 APT there? Would that allow this case be just collapsed into the case of having the fee payer set because the code would check 0x0's balance and use 2 APT?